### PR TITLE
Low-impact cross fixes

### PR DESCRIPTION
**Pull request by dtzWill**
###### Motivation for this change

Not exhaustive, but manageable chunk of fixes from #30882 (and a few local ones) that don't cause mass-rebuild.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/gcc/6/default.nix [hunk 1]
```diff
@@ -314,7 +314,44 @@ stdenv.mkDerivation ({
   + stdenv.lib.optionalString (langJava || langGo) ''
     export lib=$out;
   ''
-  ;
+  + stdenv.lib.optionalString (buildPlatform != hostPlatform)
```
Comment (Ericson2314) [inline pkgs/development/compilers/gcc/6/default.nix]: Let's skip GCC for the moment

File: pkgs/development/compilers/llvm/4/llvm.nix [hunk 2]
```diff
@@ -27,6 +28,17 @@ let
   # Used when creating a version-suffixed symlink of libLLVM.dylib
   shortVersion = with stdenv.lib;
     concatStringsSep "." (take 2 (splitString "." release_version));
+
```
Comment (Ericson2314) [inline pkgs/development/compilers/llvm/4/llvm.nix]: And also skip llvm

File: pkgs/development/libraries/libgpg-error/default.nix [hunk 3]
```diff
@@ -16,6 +16,7 @@ stdenv.mkDerivation rec {
 
   # If architecture-dependent MO files aren't available, they're generated
   # during build, so we need gettext for cross-builds.
+  crossAttrs.nativeBuildInputs = [ gettext buildPackages.stdenv.cc ];
```
Comment (Ericson2314) [inline pkgs/development/libraries/libgpg-error/default.nix]: Need to get rid of cross attrs and some other things. I can force push to fix or just skip for now.

File: pkgs/development/libraries/libssh2/default.nix [hunk 4]
```diff
@@ -29,6 +29,14 @@ stdenv.mkDerivation rec {
     '';
   };
 
+  # Hack: when cross-compiling we need to manually add rpaths to ensure that
```
Comment (Ericson2314) [inline pkgs/development/libraries/libssh2/default.nix]: No longer need this hack.

**Review by Ericson2314 — APPROVED** [review]
Good after those few things. Thanks!

File: pkgs/development/libraries/libuv/default.nix [hunk 5]
```diff
@@ -38,7 +38,7 @@ stdenv.mkDerivation rec {
   # These should be turned back on, but see https://github.com/NixOS/nixpkgs/issues/23651
   # For now the tests are just breaking large swaths of the nixpkgs binary cache for Darwin,
   # and I'd rather have everything else work at all than have stronger assurance here.
-  doCheck = !stdenv.isDarwin;
+  doCheck = !stdenv.isDarwin && stdenv.buildPlatform == stdenv.hostPlatform;
```
Comment (dezgeg) [inline pkgs/development/libraries/libuv/default.nix]: We shouldn't propagate this everywhere. Stdenv should just disable doCheck when cross compiling.

File: pkgs/tools/compression/xz/default.nix [hunk 6]
```diff
@@ -10,7 +10,7 @@ stdenv.mkDerivation rec {
 
   outputs = [ "bin" "dev" "out" "man" "doc" ];
 
-  doCheck = true;
+  doCheck = stdenv.hostPlatform == stdenv.buildPlatform;
```
Comment (dezgeg) [inline pkgs/tools/compression/xz/default.nix]: Ditto.

**Comment by dtzWill** [comment]
* Dropped gcc, llvm commits, per request.  Should they go into a separate PR?
* Dropping libgpg-error commit, I'm not sure what you'd like done.  Can you make a PR for this?
* Dropping libssh2 hack, glad it's not needed :).
* Rebased onto master since I'm force-pushing anyway

**Comment by dtzWill** [comment]
@dezgeg would you mind if the stdenv doCheck fixes, this was discussed and the consensus seems to be that this is the way to go for now: https://github.com/NixOS/nixpkgs/pull/30883

As mentioned there, cleanup would be straightforward  if we find and agree on how to better handle this.

**Comment by dezgeg** [comment]
No, I really think that pattern shouldn't spread to everywhere in nixpkgs.

**Comment by dtzWill** [comment]
Dropping perl-cross, needs some work (not being critical, but it's just not incomplete and doesn't work yet).

Thought it did work, perhaps something it needs got lost in some shuffling, will pull it out.

**Comment by Ericson2314** [comment]
Ok the permenant solution I want for `doCheck` is a mass rebuild, but I can dig up a non-mass-rebuild rebuild for now.

**Comment by dtzWill** [comment]
@Ericson2314 if you could revive that it'd be great!

**Pull request closed without merging**